File: dart_sample/lib/app_bar_leading_width_demo_page.dart
import 'package:flutter/material.dart';

import 'counter_widgets.dart';

class AppBarLeadingWidthDemoPage extends StatefulWidget {
  const AppBarLeadingWidthDemoPage({super.key});

  @override
  State<AppBarLeadingWidthDemoPage> createState() =>
      _AppBarLeadingWidthDemoPageState();
}

class _AppBarLeadingWidthDemoPageState
    extends State<AppBarLeadingWidthDemoPage> {
  double _themeLeadingWidth = 88;
  bool _useWidgetOverride = false;
  double _widgetLeadingWidth = 120;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 10,
      children: <Widget>[
        const Text(
          'AppBar theme leadingWidth',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Runtime probe for AppBar leading slot width resolution: widget override beats appBarTheme, appBarTheme beats default 56.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'theme 56',
              onTap: () => _setThemeLeadingWidth(56),
              width: 92,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: 'theme 88',
              onTap: () => _setThemeLeadingWidth(88),
              width: 92,
              background: const Color(0xFFDCE3ED),
            ),
            _buildButton(
              label: 'theme 120',
              onTap: () => _setThemeLeadingWidth(120),
              width: 102,
              background: const Color(0xFFDCE3ED),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'override off',
              onTap: () => _setUseWidgetOverride(false),
              width: 108,
              background: const Color(0xFFE9F5EC),
            ),
            _buildButton(
              label: 'override on',
              onTap: () => _setUseWidgetOverride(true),
              width: 108,
              background: const Color(0xFFE9F5EC),
            ),
          ],
        ),
        Row(
          spacing: 8,
          children: <Widget>[
            _buildButton(
              label: 'widget 72',
              onTap: () => _setWidgetLeadingWidth(72),
              width: 92,
              background: const Color(0xFFF3E8D8),
            ),
            _buildButton(
              label: 'widget 96',
              onTap: () => _setWidgetLeadingWidth(96),
              width: 92,
              background: const Color(0xFFF3E8D8),
            ),
            _buildButton(
              label: 'widget 120',
              onTap: () => _setWidgetLeadingWidth(120),
              width: 102,
              background: const Color(0xFFF3E8D8),
            ),
          ],
        ),
        Text(
          'themeLeadingWidth=${_themeLeadingWidth.toStringAsFixed(0)}, override=${_useWidgetOverride ? 'on' : 'off'}, widgetLeadingWidth=${_widgetLeadingWidth.toStringAsFixed(0)}, effective=${_effectiveLeadingWidthLabel()}',
          style: const TextStyle(fontSize: 12, color: Colors.blueGrey),
        ),
        Container(
          width: 300,
          color: const Color(0xFFE7EDF6),
          padding: const EdgeInsets.all(10),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            spacing: 8,
            children: <Widget>[
              const Text(
                'Themed app bar preview',
                style: TextStyle(fontSize: 11, color: Colors.black54),
              ),
              _buildThemedPreview(),
              const Text(
                'Default app bar reference (leadingWidth = 56)',
                style: TextStyle(fontSize: 11, color: Colors.black54),
              ),
              _buildDefaultReferencePreview(),
            ],
          ),
        ),
      ],
    );
  }

  Widget _buildThemedPreview() {
    return Theme(
      data: ThemeData.light().copyWith(
        appBarTheme: AppBarTheme(
          backgroundColor: const Color(0xFF1E3A5F),
          foregroundColor: Colors.white,
          leadingWidth: _themeLeadingWidth,
        ),
      ),
      child: AppBar(
        title: const Text('Theme leading width'),
        leadingWidth: _useWidgetOverride ? _widgetLeadingWidth : null,
        leading: _buildLeadingProbe(),
        actions: const <Widget>[Text('A1', style: TextStyle(fontSize: 11))],
      ),
    );
  }

  Widget _buildDefaultReferencePreview() {
    return AppBar(
      title: const Text('Default leading width'),
      backgroundColor: const Color(0xFF1E3A5F),
      foregroundColor: Colors.white,
      leading: _buildLeadingProbe(),
      actions: const <Widget>[Text('A1', style: TextStyle(fontSize: 11))],
    );
  }

  static Widget _buildLeadingProbe() {
    return Container(
      width: 24,
      height: 24,
      color: const Color(0xFFFFB703),
      child: const Center(
        child: Text('L', style: TextStyle(fontSize: 11, color: Colors.black)),
      ),
    );
  }

  Widget _buildButton({
    required String label,
    required VoidCallback onTap,
    required double width,
    required Color background,
  }) {
    return SizedBox(
      width: width,
      child: CounterTapButton(
        label: label,
        onTap: onTap,
        background: background,
        foreground: Colors.black,
        fontSize: 12,
        padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 8),
      ),
    );
  }

  void _setThemeLeadingWidth(double value) {
    setState(() {
      _themeLeadingWidth = value;
    });
  }

  void _setUseWidgetOverride(bool value) {
    setState(() {
      _useWidgetOverride = value;
    });
  }

  void _setWidgetLeadingWidth(double value) {
    setState(() {
      _widgetLeadingWidth = value;
    });
  }

  String _effectiveLeadingWidthLabel() {
    final double effective = _useWidgetOverride
        ? _widgetLeadingWidth
        : _themeLeadingWidth;
    return effective.toStringAsFixed(0);
  }
}

File: src/Sample/Flutter.Net/AppBarLeadingWidthDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Flutter.Material;
using Flutter.Rendering;
using Flutter.Widgets;

// Dart parity source (reference): dart_sample/lib/app_bar_leading_width_demo_page.dart (exact sample parity)

namespace Flutter.Net;

public sealed class AppBarLeadingWidthDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new AppBarLeadingWidthDemoPageState();
    }
}

internal sealed class AppBarLeadingWidthDemoPageState : State
{
    private double _themeLeadingWidth = 88;
    private bool _useWidgetOverride;
    private double _widgetLeadingWidth = 120;

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 10,
            children:
            [
                new Text("AppBar theme leadingWidth", fontSize: 20, color: Colors.Black),
                new Text(
                    "Runtime probe for AppBar leading slot width resolution: widget override beats appBarTheme, appBarTheme beats default 56.",
                    fontSize: 14,
                    color: Colors.DimGray),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("theme 56", () => SetThemeLeadingWidth(56), width: 92, colorHex: "#FFDCE3ED"),
                        BuildButton("theme 88", () => SetThemeLeadingWidth(88), width: 92, colorHex: "#FFDCE3ED"),
                        BuildButton("theme 120", () => SetThemeLeadingWidth(120), width: 102, colorHex: "#FFDCE3ED"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("override off", () => SetUseWidgetOverride(false), width: 108, colorHex: "#FFE9F5EC"),
                        BuildButton("override on", () => SetUseWidgetOverride(true), width: 108, colorHex: "#FFE9F5EC"),
                    ]),
                new Row(
                    spacing: 8,
                    children:
                    [
                        BuildButton("widget 72", () => SetWidgetLeadingWidth(72), width: 92, colorHex: "#FFF3E8D8"),
                        BuildButton("widget 96", () => SetWidgetLeadingWidth(96), width: 92, colorHex: "#FFF3E8D8"),
                        BuildButton("widget 120", () => SetWidgetLeadingWidth(120), width: 102, colorHex: "#FFF3E8D8"),
                    ]),
                new Text(
                    $"themeLeadingWidth={_themeLeadingWidth:0}, override={(_useWidgetOverride ? "on" : "off")}, widgetLeadingWidth={_widgetLeadingWidth:0}, effective={EffectiveLeadingWidthLabel()}",
                    fontSize: 12,
                    color: Colors.DarkSlateGray),
                new Container(
                    width: 300,
                    color: Color.Parse("#FFE7EDF6"),
                    padding: new Thickness(10),
                    child: new Column(
                        crossAxisAlignment: CrossAxisAlignment.Stretch,
                        spacing: 8,
                        children:
                        [
                            new Text("Themed app bar preview", fontSize: 11, color: Colors.DimGray),
                            BuildThemedPreview(),
                            new Text("Default app bar reference (leadingWidth = 56)", fontSize: 11, color: Colors.DimGray),
                            BuildDefaultReferencePreview(),
                        ])),
            ]);
    }

    private Widget BuildThemedPreview()
    {
        return new Theme(
            data: ThemeData.Light with
            {
                AppBarTheme = new AppBarThemeData(
                    BackgroundColor: Color.Parse("#FF1E3A5F"),
                    ForegroundColor: Colors.White,
                    LeadingWidth: _themeLeadingWidth),
            },
            child: new AppBar(
                titleText: "Theme leading width",
                leadingWidth: _useWidgetOverride ? _widgetLeadingWidth : null,
                leading: BuildLeadingProbe(),
                actions:
                [
                    new Text("A1", fontSize: 11),
                ]));
    }

    private Widget BuildDefaultReferencePreview()
    {
        return new AppBar(
            titleText: "Default leading width",
            backgroundColor: Color.Parse("#FF1E3A5F"),
            foregroundColor: Colors.White,
            leading: BuildLeadingProbe(),
            actions:
            [
                new Text("A1", fontSize: 11),
            ]);
    }

    private static Widget BuildLeadingProbe()
    {
        return new Container(
            width: 24,
            height: 24,
            color: Color.Parse("#FFFFB703"),
            child: new Center(
                child: new Text("L", fontSize: 11, color: Colors.Black)));
    }

    private Widget BuildButton(string label, Action onTap, double width, string colorHex)
    {
        return new SizedBox(
            width: width,
            child: new CounterTapButton(
                label: label,
                onTap: onTap,
                background: Color.Parse(colorHex),
                foreground: Colors.Black,
                fontSize: 12,
                padding: new Thickness(10, 8)));
    }

    private void SetThemeLeadingWidth(double value)
    {
        SetState(() => _themeLeadingWidth = value);
    }

    private void SetUseWidgetOverride(bool value)
    {
        SetState(() => _useWidgetOverride = value);
    }

    private void SetWidgetLeadingWidth(double value)
    {
        SetState(() => _widgetLeadingWidth = value);
    }

    private string EffectiveLeadingWidthLabel()
    {
        return (_useWidgetOverride ? _widgetLeadingWidth : _themeLeadingWidth).ToString("0");
    }
}
